pallets/pallet-bonded-coins/src/curves/mod.rs

```diff
@@ -111,6 +111,31 @@ pub(crate) fn convert_to_fixed<T: Config>(
 	let decimals = 10u128
 		.checked_pow(u32::from(denomination))
 		.ok_or(ArithmeticError::Overflow)?;
-	let scaled_x = x.checked_div(decimals).ok_or(ArithmeticError::DivisionByZero)?;
-	scaled_x.checked_to_fixed().ok_or(ArithmeticError::Overflow)
+
+	// Divide `x` to separate the quotient and remainder
+	let quotient = x.checked_div(decimals).ok_or(ArithmeticError::DivisionByZero)?;
+	let remainder = x.checked_rem_euclid(decimals).ok_or(ArithmeticError::DivisionByZero)?;
```

Comment:
To avoid truncating amounts that are smaller than a multiple of the denomination, we need to convert to fixed first, then divide by the denomination. This of course can easily result in overflow, which defies the purpose of scaling by denomination.
This is why I'm adopting a combined strategy: On one leg, perform scaling as before, truncating fractionals, and on the other leg, take the remainder of the integer division, convert it to fixed, then divide by the denomination. This gives us the fractional part that was previously dropped, which we can now add to the result.

Reply:
I kinda have the feeling, this is a bit overengineered: maybe I am wrong. The additional bits gained by the second step, are anyway dropped in the combination step, if the denomination is bigger than the fractional part of the Float. 

> This of course can easily result in overflow

That's right, but also quite an edge case, right? We kinda have some numeric bounds. A user can also provide an amount, which would not fit into an u128, then we have the same problem. I would argue the possible values are big enough

Reply:
> I kinda have the feeling, this is a bit overengineered: maybe I am wrong.

What do you not agree with, the problem analysis or the proposed solution? 

__On the problem:__
I don't actually think we disagree here, but just to make sure: I do think it needs to be addressed, because it effectively means you cannot use sub-units at all; everything is rounded down. You cannot mint 1.4 xyz-coins - you'd be minting 1. You cannot burn 0.5 - this would just error. If there is any supply on any of the currencies that's below the denomination, it is going to be completely ignored in the calculations.

If that is what we want, _why even have a denomination at all_? If units are non-divisible, we may just assume a denomination of 0 everywhere. This would avoid a number of nasty bugs, for example the fact that you currently pay for the rounded down amount (e.g., 1 xyz), but what's minted is the exact amount you specified (1.4 xyz).

__On the solution:__
Is your preference simply to just convert to fixed first, then divide by the denomination, and accept the potential to overflow? Are you saying that trying to solve the overflow issue is overkill? 
In that case, we could handle amounts of up to ~1.8e+22. For a denomination of 10 that seems enough(~1.8 Trillion), for denomination of 15 that would mean we'd overflow at around 18 Million.

> I would argue the possible values are big enough

Possibly, I wouldn't really know how to make a judgement on this. If our balances would be u64 values, certainly; given that they are u128 we'd be wasting a ton of space that we can't effectively make use of. 

Still, if we accept these limits, we somehow need to enforce them, meaning that no bonded currency can have a supply > 2^74 (capacity of the fixed type we're using). Right now there is nothing preventing you from minting coins until the pool is past this limit, after which every attempt to mint or burn will fail and it's stuck completely. I wouldn't call that an edge case, but rather a nasty-ass bug.

>  A user can also provide an amount, which would not fit into an u128, then we have the same problem.

They actually can't; any transaction where the `amount_to_{mint,burn}` parameter is not exactly 128 bits long is invalid and will never be included in a block.

> The additional bits gained by the second step, are anyway dropped in the combination step, if the denomination is bigger than the fractional part of the Float.

I don't understand this at all, can you elaborate? And you mean Fixed, right? Right now we're effectively working with integers, and the fractional part is always 0, with my proposed there would be a non-zero fractional part, so I don't exactly get how these "extra bits are being dropped"?

A quick example is this test case:

```rust
 #[test]
 fn test_convert_to_fixed_with_remainder() {
 	let x = 1050u128;
 	let denomination = 2u8; // 10^2 = 100

 	let result = convert_to_fixed::<Test>(x, denomination).unwrap();
 	let expected = Float::from_num(10.5); // 1050 / 100 = 10.5

 	assert_eq!(result, expected);
 }
```

Before my change, this would error, as the function returned `10`; now it returns the correct result, `10.5`.

Reply:
**On the problem**

I totally agree with you. This is a bug which we have to address. 0.5 UNITs have to be representable, which was also the case in the previous implementation (I hope). No discussion. 

**On the Solution** 

We should address all overflows. But I personally think, it is enough, to mint 18m coins for a denomination of 15 in a single step - if a user wants to mint more, he needs to provide a smaller denomination. The selection of the denomination should be an active decision, based on the user's need. Currently, nothing is stopping the user to provide a denomination of 100. It does not really make sense, but he can do that. 


> If our balances would be u64 values, certainly; given that they are u128 we'd be wasting a ton of space that we can't effectively make use of.

I don't really get this point.

> after which every attempt to mint or burn will fail and it's stuck completely.

That's not true. Every coin minted is also burnable. The user is not stuck, he just hit the numeric bounds for his pool, which he has to consider in the pool creation. 

Maybe instead of calculating the normalized amount of an u128, the tx should directly take the normalized amount. Then we have no issues, we are faster, and the user really gets what he wants (no rounding at all).

Reply:
> I don't understand this at all, can you elaborate? And you mean Fixed, right? 

Yes, I mean fixed. By the current Float Implementation, we have 16 digits for the frac part. If the denomination is 17 for instance, we would drop the LSB

Reply:
> 0.5 UNITs have to be representable, which was also the case in the previous implementation (I hope).

They weren't, that's why this PR exists. I couldn't get unit tests to pass because of that. 


> We should address all overflows. But I personally think, it is enough, to mint 18m coins for a denomination of 15 in a single step - if a user wants to mint more, he needs to provide a smaller denomination.

Agree, but unfortunately that's not the real problem we're facing (see below)


> > If our balances would be u64 values, certainly; given that they are u128 we'd be wasting a ton of space that we can't effectively make use of.

The supply that any currency may have before shit hits the fan is capped at 2^74 (see below). So we have 74 useable bits out of 128. A u64 balance is closer to what we can use, and would also naturally prevent balances to move into value ranges that break everything. 



> Every coin minted is also burnable. The user is not stuck, he just hit the numeric bounds for his pool, which he has to consider in the pool creation. 

Take a look at https://github.com/KILTprotocol/kilt-node/blob/pallet_bonded_coins/pallets/pallet-bonded-coins/src/lib.rs#L834. This is invoke during mint on the balance before mint, but not after. 
The result: the balance can increase to the point where this function can't handle it anymore, but the effects will only be seen on the next burn/mint, which will then fail regardless of the amount selected.  

> Maybe instead of calculating the normalized amount of an u128, the tx should directly take the normalized amount. Then we have no issues, we are faster, and the user really gets what he wants (no rounding at all). 

That's giving up on denominations. If there is not way to mint or burn 0.5 units, why even represent them?